test/distributed/org/apache/cassandra/distributed/impl/MessageFilters.java

```diff
@@ -86,65 +69,77 @@ public boolean equals(Object that)
 
         public boolean equals(Filter that)
         {
-            return Arrays.equals(from, that.from)
-                    && Arrays.equals(to, that.to)
-                    && Arrays.equals(verbs, that.verbs);
+            return that.matchers.equals(matchers);
         }
 
-        public boolean matches(int from, int to, int verb)
-        {
-            return (this.from == null || Arrays.binarySearch(this.from, from) >= 0)
-                    && (this.to == null || Arrays.binarySearch(this.to, to) >= 0)
-                    && (this.verbs == null || Arrays.binarySearch(this.verbs, verb) >= 0);
-        }
-
-        public Filter restore()
+        public Filter off()
         {
             filters.remove(this);
```

Comment:
was this working before?  If I am reading this properly, `this` is an instance of `Filter` which is a composite of other `IMatcher`.  If I disable the composite it attempts to remove itself from itself?  

ill need to look closer, but this wasn't changed in this so ignoring for now.

Reply:
So `off` on a filter chain means exactly what it does: removes and disables this specific filter chain. I've added a test to validate the logic, too.